File: ios/Showcase/Targets/AppUIKit/Sources/walletSettings/WalletSettingsHomeView.swift
import DeviceCheck
import SpruceIDMobileSdk
import SpruceIDMobileSdkRs
import SwiftUI

struct WalletSettingsHome: Hashable {}

struct WalletSettingsHomeView: View {
    @Binding var path: NavigationPath

    func onBack() {
        while !path.isEmpty {
            path.removeLast()
        }
    }

    var body: some View {
        VStack {
            WalletSettingsHomeHeader(onBack: onBack)
            WalletSettingsHomeBody(
                path: $path,
                onBack: onBack
            )
        }
        .navigationBarBackButtonHidden(true)
    }
}

struct WalletSettingsHomeHeader: View {
    var onBack: () -> Void

    var body: some View {
        HStack {
            Text("Preferences")
                .font(.customFont(font: .inter, style: .bold, size: .h2))
                .padding(.leading, 30)
                .foregroundStyle(Color("ColorStone950"))
            Spacer()
            Button {
                onBack()
            } label: {
                ZStack {
                    RoundedRectangle(cornerRadius: 8)
                        .foregroundColor(Color("ColorStone950"))
                        .frame(width: 36, height: 36)
                    Image("User")
                        .foregroundColor(Color("ColorStone50"))
                }
            }
            .padding(.trailing, 20)
        }
        .padding(.top, 10)
    }
}

struct WalletSettingsHomeBody: View {
    @EnvironmentObject private var credentialPackObservable:
        CredentialPackObservable
    @EnvironmentObject private var hacApplicationObservable:
        HacApplicationObservable
    @StateObject private var environmentConfig = EnvironmentConfig.shared
    @Binding var path: NavigationPath
    var onBack: () -> Void
    @State private var isApplyingForMdl = false

    @ViewBuilder
    var activityLogButton: some View {
        Button {
            path.append(WalletSettingsActivityLog())
        } label: {
            SettingsHomeItem(
                image: "List",
                title: "Activity Log",
                description: "View and export activity history"
            )
        }
    }

    @ViewBuilder
    var deleteAllCredentials: some View {
        Button {
            Task {
                do {
                    let credentialPacks = credentialPackObservable
                        .credentialPacks
                    try await credentialPacks.asyncForEach { credentialPack in
                        try await credentialPackObservable.delete(
                            credentialPack: credentialPack
                        )
                        credentialPack.list().forEach { credential in
                            let credentialInfo = getCredentialIdTitleAndIssuer(
                                credentialPack: credentialPack,
                                credential: credential
                            )
                            _ = WalletActivityLogDataStore.shared.insert(
                                credentialPackId: credentialPack.id.uuidString,
                                credentialId: credentialInfo.0,
                                credentialTitle: credentialInfo.1,
                                issuer: credentialInfo.2,
                                action: "Deleted",
                                dateTime: Date(),
                                additionalInformation: ""
                            )
                        }
                    }
                    let _ = HacApplicationDataStore.shared.deleteAll()
                } catch {
                    // TODO: display error message
                    print(error)
                }
            }
        } label: {
            Text("Delete all added credentials")
                .frame(width: UIScreen.screenWidth)
                .padding(.horizontal, -20)
                .font(.customFont(font: .inter, style: .medium, size: .h4))
        }
        .foregroundColor(.white)
        .padding(.vertical, 13)
        .background(Color("ColorRose700"))
        .cornerRadius(8)
    }

    @ViewBuilder
    var generateMockMdlButton: some View {
        Button {
            Task {
                await generateMockMdl()
            }
        } label: {
            SettingsHomeItem(
                image: "GenerateMockMdl",
                title: "Generate mDL",
                description:
                    "Generate a fresh test mDL issued by the SpruceID Test CA"
            )
        }
    }

    @ViewBuilder
    var applyForSpruceMdlButton: some View {
        Button {
            isApplyingForMdl = true
            Task {
                do {
                    let walletAttestation =
                        try await hacApplicationObservable
                        .getWalletAttestation()
                        .unwrap()

                    let issuance =
                        try await hacApplicationObservable.issuanceClient
                        .newIssuance(walletAttestation: walletAttestation)

                    let hacApplication = HacApplicationDataStore.shared.insert(
                        issuanceId: issuance
                    )

                    let status = try await hacApplicationObservable.issuanceClient.checkStatus(
                        issuanceId: issuance,
                        walletAttestation: walletAttestation
                    )

                    switch status {
                        case .proofingRequired(let proofingUrl):
                            if let hacApplication = hacApplication {
                                if let url = URL(string: proofingUrl) {
                                    UIApplication.shared.open(
                                        url,
                                        options: [:],
                                        completionHandler: nil
                                    )
                                } else {
                                    print("Invalid proofing URL")
                                }
                            } else {
                                print("hacApplication is nil")
                            }

                        case .readyToProvision(_):
                            print("Expected ProofingRequired status")
                            ToastManager.shared.showError(
                                message:
                                    "Error during attestation: Expected ProofingRequired status"
                            )
                        case .awaitingManualReview:
                            print("Expected ProofingRequired status")
                            ToastManager.shared.showError(
                                message:
                                    "Error during attestation: Expected ProofingRequired status"
                            )
                        case .applicationDenied:
                            print("Expected ProofingRequired status")
                            ToastManager.shared.showError(
                                message:
                                    "Error during attestation: Expected ProofingRequired status"
                            )
                        }

                } catch let error as DCError {
                    ToastManager.shared.showError(
                        message:
                            "App Attestation failed: \(error.localizedDescription)"
                    )
                } catch {
                    ToastManager.shared.showError(
                        message:
                            "Error during attestation: \(error.localizedDescription)"
                    )
                }
                isApplyingForMdl = false
            }
        } label: {
            SettingsHomeItem(
                image: "ApplySpruceMdl",
                title: "Apply for Spruce mDL",
                description:
                    "Verify your identity in order to claim this high assurance credential"
            )
        }
        .disabled(isApplyingForMdl)
        .opacity(isApplyingForMdl ? 0.5 : 1.0)
    }

    @ViewBuilder
    var devModeButton: some View {
        Button {
            environmentConfig.toggleDevMode()
        } label: {
            SettingsHomeItem(
                image: "DevMode",
                title:
                    "\(environmentConfig.isDevMode ? "Disable" : "Enable") Dev Mode",
                description:
                    "Warning: Dev mode will use in development services and is not recommended for production use"
            )
        }
    }

    var body: some View {
        VStack {
            VStack {
                activityLogButton
                generateMockMdlButton
                applyForSpruceMdlButton
                devModeButton
                Spacer()
                deleteAllCredentials
            }
        }
        .padding(.vertical, 20)
        .padding(.horizontal, 30)
    }
}

File: android/Showcase/src/main/java/com/spruceid/mobilesdkexample/walletsettings/WalletSettingsHomeView.kt
package com.spruceid.mobilesdkexample.walletsettings

import android.content.Intent
import android.net.Uri
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.spruceid.mobile.sdk.rs.FlowState
import com.spruceid.mobilesdkexample.R
import com.spruceid.mobilesdkexample.config.EnvironmentConfig
import com.spruceid.mobilesdkexample.db.HacApplications
import com.spruceid.mobilesdkexample.db.WalletActivityLogs
import com.spruceid.mobilesdkexample.navigation.Screen
import com.spruceid.mobilesdkexample.ui.theme.ColorRose600
import com.spruceid.mobilesdkexample.ui.theme.ColorStone50
import com.spruceid.mobilesdkexample.ui.theme.ColorStone950
import com.spruceid.mobilesdkexample.ui.theme.Inter
import com.spruceid.mobilesdkexample.utils.ControlledSimpleDeleteAlertDialog
import com.spruceid.mobilesdkexample.utils.SettingsHomeItem
import com.spruceid.mobilesdkexample.utils.activityHiltViewModel
import com.spruceid.mobilesdkexample.utils.getCredentialIdTitleAndIssuer
import com.spruceid.mobilesdkexample.utils.getCurrentSqlDate
import com.spruceid.mobilesdkexample.viewmodels.CredentialPacksViewModel
import com.spruceid.mobilesdkexample.viewmodels.HacApplicationsViewModel
import com.spruceid.mobilesdkexample.viewmodels.WalletActivityLogsViewModel
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

@Composable
fun WalletSettingsHomeView(
    navController: NavController
) {
    Column(
        Modifier
            .padding(all = 20.dp)
            .padding(top = 20.dp)
    ) {
        WalletSettingsHomeHeader(
            onBack = {
                navController.navigate(
                    Screen.HomeScreen.route.replace("{tab}", "wallet")
                ) {
                    popUpTo(0)
                }
            }
        )
        WalletSettingsHomeBody(
            navController = navController
        )
    }
}

@Composable
fun WalletSettingsHomeHeader(onBack: () -> Unit) {
    Row(verticalAlignment = Alignment.CenterVertically) {
        Text(
            text = "Preferences",
            fontFamily = Inter,
            fontWeight = FontWeight.SemiBold,
            fontSize = 20.sp,
            color = ColorStone950
        )
        Spacer(Modifier.weight(1f))
        Box(
            contentAlignment = Alignment.Center,
            modifier = Modifier
                .width(36.dp)
                .height(36.dp)
                .padding(start = 4.dp)
                .clip(shape = RoundedCornerShape(8.dp))
                .background(ColorStone950)
                .clickable {
                    onBack()
                }
        ) {
            Image(
                painter = painterResource(id = R.drawable.user),
                contentDescription = stringResource(id = R.string.user),
                colorFilter = ColorFilter.tint(ColorStone50),
                modifier = Modifier
                    .width(20.dp)
                    .height(20.dp)
            )
        }
    }
}

@Composable
fun WalletSettingsHomeBody(
    navController: NavController
) {
    val credentialPacksViewModel: CredentialPacksViewModel = activityHiltViewModel()
    val walletActivityLogsViewModel: WalletActivityLogsViewModel = activityHiltViewModel()
    val hacApplicationsViewModel: HacApplicationsViewModel = activityHiltViewModel()
    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    var isApplyingForMdl by remember { mutableStateOf(false) }
    var showDeleteDialog by remember { mutableStateOf(false) }
    val isDevMode by EnvironmentConfig.isDevMode.collectAsState()

    Column(
        Modifier
            .padding(top = 10.dp)
            .navigationBarsPadding(),
    ) {
        SettingsHomeItem(
            icon = {
                Image(
                    painter = painterResource(id = R.drawable.verification_activity_log),
                    contentDescription = stringResource(id = R.string.verification_activity_log),
                    modifier = Modifier.padding(end = 5.dp),
                )
            },
            name = "Activity Log",
            description = "View and export activity history",
            action = {
                navController.navigate(Screen.WalletSettingsActivityLogScreen.route)
            }
        )

        GenerateMockMdlButton(credentialPacksViewModel = credentialPacksViewModel, walletActivityLogsViewModel = walletActivityLogsViewModel)

        SettingsHomeItem(
            icon = {
                Image(
                    painter = painterResource(id = R.drawable.apply_spruceid_mdl),
                    contentDescription = stringResource(id = R.string.apply_spruceid_mdl),
                    modifier = Modifier.padding(end = 5.dp),
                )
            },
            name = "Apply for Spruce mDL",
            description = "Verify your identity in order to claim this high assurance credential",
            enabled = !isApplyingForMdl,
            action = {
                scope.launch {
                    isApplyingForMdl = true
                    try {
                        val walletAttestation = hacApplicationsViewModel.getWalletAttestation()
                        walletAttestation?.let {
                            val issuance =
                                hacApplicationsViewModel.issuanceClient
                                    .newIssuance(walletAttestation)

                            val hacApplication = hacApplicationsViewModel.saveApplication(
                                HacApplications(issuanceId = issuance)
                            )
                            val status = hacApplicationsViewModel.issuanceClient.checkStatus(
                                issuance,
                                walletAttestation
                            )

                            when (status) {
                                is FlowState.ProofingRequired -> {
                                    val intent = Intent(
                                        Intent.ACTION_VIEW,
                                        Uri.parse(status.proofingUrl)
                                    )
                                    context.startActivity(intent)
                                }

                                is FlowState.ReadyToProvision -> {
                                    print("Issuance started with invalid state, please check.")
                                }

                                is FlowState.ApplicationDenied -> {
                                    print("Issuance started with invalid state, please check.")
                                }

                                is FlowState.AwaitingManualReview -> {
                                    print("Issuance started with invalid state, please check.")
                                }
                            }
                        }
                    } finally {
                        isApplyingForMdl = false
                    }
                }
            }
        )

        SettingsHomeItem(
            icon = {
                Image(
                    painter = painterResource(id = R.drawable.dev_mode),
                    contentDescription = stringResource(id = R.string.dev_mode),
                    modifier = Modifier.padding(end = 5.dp),
                )
            },
            name = "${if (isDevMode) "Disable" else "Enable"} Dev Mode",
            description = "Warning: Dev mode will use in development services and is not recommended for production use",
            action = {
                EnvironmentConfig.toggleDevMode()
            }
        )

        Spacer(Modifier.weight(1f))
        Button(
            onClick = {
                showDeleteDialog = true
            },
            shape = RoundedCornerShape(5.dp),
            colors = ButtonDefaults.buttonColors(
                containerColor = ColorRose600,
                contentColor = Color.White,
            ),
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 30.dp)
        ) {
            Text(
                text = "Delete all added credentials",
                fontFamily = Inter,
                fontWeight = FontWeight.SemiBold,
                color = Color.White,
            )
        }
        ControlledSimpleDeleteAlertDialog(
            showDialog = showDeleteDialog,
            message = "Are you sure you want to delete all the credentials? This action cannot be undone.",
            onConfirm = {
                GlobalScope.launch {
                    credentialPacksViewModel.deleteAllCredentialPacks(onDeleteCredentialPack = { credentialPack ->
                        credentialPack.list().forEach { credential ->
                            val credentialInfo =
                                getCredentialIdTitleAndIssuer(
                                    credentialPack,
                                    credential
                                )
                            walletActivityLogsViewModel.saveWalletActivityLog(
                                walletActivityLogs = WalletActivityLogs(
                                    credentialPackId = credentialPack.id().toString(),
                                    credentialId = credentialInfo.first,
                                    credentialTitle = credentialInfo.second,
                                    issuer = credentialInfo.third,
                                    action = "Deleted",
                                    dateTime = getCurrentSqlDate(),
                                    additionalInformation = ""
                                )
                            )
                        }
                    })
                    hacApplicationsViewModel.deleteAllApplications()
                    showDeleteDialog = false
                }
            },
            onClose = {
                showDeleteDialog = false
            },
            confirmButtonText = "Delete"
        )

    }
}
